### ISSUE TITLE
message box should not accept "anything" messages

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#1491]
message boxes currently accept "anything" messages. They basically ignore the selector and pass the remaining arguments like a list message.

I think this is a problem because message box already accepts messages like `[set(`, `[addcomma(`, `[addsemi(`, etc.

Generally, Pd objects that accept custom methods should *not* accept "anythings" because it's not possible to distinguish between the two. This can lead to two problems:
* the user might accidentally send a message with a "reserved" selector and get unexpected results
* we can't add new methods because users might already have patches which use these symbols

For example, that's the reason why `[list store]`, contrary to `[list]` doesn't accept anythings.

---

TL;DR: I would suggest to remove the "anything" method from message box. I doubt that there a patches which rely on it, but they would get a clear error message in the console anyway ("no method for...")

**Comment by porres** [comment]
but all that happens is that the message box gets activated and passes its contents, right?

**Comment by Spacechild1** [comment]
See the two problems I've pointed out above.

**Comment by porres** [comment]
oh, you mean when dealing with "$1", I get it now that I read the email on the pd list

**Comment by Spacechild1** [comment]
I clarified the PR description. Thanks!

**Comment by porres** [comment]
I remember encountering bumps and having to turn something into a symbol so I could put it in a "$1" inside a message.

I don't remember turning to 'list' to use "$1"/"$2" to get elements from 'anythings', or getting these elements without turning them to 'list', though...

**Comment by porres** [comment]
btw, I've recently documented how this works

<img width="527" alt="Screen Shot 2021-12-02 at 11 25 33" src="https://user-images.githubusercontent.com/5360142/144440891-81bf9eea-2250-423d-b27a-3e8344e5465b.png">

**Comment by Spacechild1** [comment]
> I remember encountering bumps and having to turn something into a symbol so I could put it in a "$1" inside a message.

This is, in fact, the same problem! `[foo(` is a Pd message with no arguments. Message box will strip the selector (= "foo") and pass the remaining arguments (= none).

**Comment by Spacechild1** [comment]
> btw, I've recently documented how this works

Oh! We definitely shouldn't advertize this as a feature. Maybe document it as unexpected behavior that will disappear in the future?

**Comment by Ant1r** [comment]
> We definitely shouldn't advertize this as a feature

too late for me... I'm sure I've used this 'trick' numerous times (I didn't know it was bad ;-)
So, while I totally understand your arguments for disallowing "anything", I'm (almost) sure it would break many of my patches...

OTOH, if the error is 'traceable' (i.e ctrl-clickable from the Pd console) then it will be possible to locate them then fix.

**Comment by Spacechild1** [comment]
> OTOH, if the error is 'traceable' (i.e ctrl-clickable from the Pd console) then it will be possible to locate them then fix.

Yes! I guess, we would also add a compatibility setting so that old patches would still work.

**Comment by porres** [comment]
i don't care that much about changing behaviour, but yeah, I wasn't explicit in my earlier messages, this definitely has the potential to break stuff and would need a compatibility flag, which seems like a last resource, in cases like the bob~ outlet just now. The output was simply wrong, it was a **bug**!

This is more of a change  motivated by consistency, better design principles, code elegancy.

It has been noted in the documentation of messages for ages now that "Anytime the message box receives any message at all, the messages in the box are all sent to their destinations." I know there are the exeptions ('set'/'etc').

This might be the case to just document in a sane way how this thing works. We should join @millerpuckette  into the discussion, since the next release already has merged new content from my documentation branch. I was just improving with more examples, but we may need to revise now how we modify this document.

**Comment by porres** [comment]
My new revision of the documentation gives more details, but doesn't explicitly document or give an example that you should get message elements from "anythings" like before. I'm not saying you shouldn't do it, neither that you should. It's kind of ambiguous, abut you can deduce it though.

If we decide we really want to keep things as is I can just explicitly document things like before.
![Uploading Screen Shot 2021-12-02 at 14.53.06.png…]()

**Comment by porres** [comment]
My new revision of the documentation gives more details, but doesn't explicitly document or give an example that you should get message elements from "anythings" like before. I'm not saying you shouldn't do it, neither that you should. It's kind of ambiguous, abut you can deduce it though.

If we decide we really want to keep things as is I can just explicitly document things like before.

<img width="857" alt="Screen Shot 2021-12-02 at 14 53 06" src="https://user-images.githubusercontent.com/5360142/144477686-0532de9d-cacd-4ba8-b3ff-a004035cae2a.png">
